plugins/webhooks/index.js

```diff
@@ -640,40 +689,60 @@ exports.register = (server, options, next) => {
 
                     request.log(['webhook', hookId], `Received event type ${type}`);
 
+                    // skipping checks
                     if (/\[(skip ci|ci skip)\]/.test(parsed.lastCommitMessage)) {
                         skipMessage = 'Skipping due to the commit message: [skip ci]';
                     }
-
                     if (ignoreUser && ignoreUser.includes(username)) {
                         message = `Skipping because user ${username} is ignored`;
                         request.log(['webhook', hookId], message);
 
                         return reply({ message }).code(204);
                     }
 
-                    await promiseToWait(WAIT_FOR_CHANGEDFILES);
                     const token = await obtainScmToken(
                         pluginOptions, userFactory, username, scmContext);
 
-                    parsed.changedFiles = await scm.getChangedFiles({
-                        payload: request.payload,
-                        type,
-                        token,
-                        scmContext
-                    });
+                    if (!parsed.sha) {
+                        try {
+                            parsed.sha = await getCommitRefSha({
+                                scm,
+                                token,
+                                ref,
+                                checkoutUrl,
+                                scmContext
+                            });
+                        } catch (err) {
+                            request.log(['webhook', hookId, 'getCommitRefSha'], err);
+
+                            // there is a possibility of scm.getCommitRefSha() is not implemented yet
+                            return reply({ message }).code(204);
+                        }
+                    }
 
-                    request.log(['webhook', hookId], `Changed files are ${parsed.changedFiles}`);
+                    if (action !== 'release' && action !== 'tag') {
+                        await promiseToWait(WAIT_FOR_CHANGEDFILES);
+
+                        parsed.changedFiles = await scm.getChangedFiles({
+                            payload: request.payload,
+                            type,
+                            token,
+                            scmContext
+                        });
+                        request.log(['webhook', hookId],
+                            `Changed files are ${parsed.changedFiles}`);
+                    }
 
                     if (type === 'pr') {
                         // disregard skip ci for pull request events
-                        return pullRequestEvent(pluginOptions, request, reply, parsed);
+                        return pullRequestEvent(pluginOptions, request, reply, parsed, token);
                     }
 
-                    return pushEvent(pluginOptions, request, reply, parsed, skipMessage);
+                    return pushEvent(pluginOptions, request, reply, parsed, skipMessage, token);
```

Comment:
We should use one `if ... else ...`

Reply:
It conflicts with `no-else-return` rule of eslint.